src/foam/box/socket/SocketConnectionBoxManager.js

```diff
@@ -0,0 +1,218 @@
+
+  properties: [
+    {
+      documentation: 'So not to block server shutdown, have sockets timeout. Catch and continue on SocketTimeoutException.',
+      class: 'Int',
+      name: 'soTimeout',
+      value: 60000
+    },
+    {
+      documentation: 'Time to wait on initial creation for connection to be established',
+      class: 'Int',
+      name: 'connectTimeout',
+      value: 60000
+    },
+    {
+      class: 'Map',
+      name: 'boxes',
+      javaFactory: `
+        return java.util.Collections.synchronizedMap(new java.util.HashMap());
+      `
+    },
+    {
+      name: 'logger',
+      class: 'FObjectProperty',
+      of: 'foam.nanos.logger.Logger',
+      visibility: 'HIDDEN',
+      javaFactory: `
+        return new PrefixLogger(new Object[] {
+          this.getClass().getSimpleName()
+        }, (Logger) getX().get("logger"));
+      `
+    }
+  ],
+
+  methods: [
+    {
+      name: 'add',
+      args: [
+        {
+          name: 'box',
+          type: 'SocketConnectionBox'
+        }
+      ],
+      javaCode: `
+      getBoxes().put(makeKey(box.getHost(), box.getPort()), box);
+      `
+    },
+    {
+      name: 'remove',
+      args: [
+        {
+          name: 'box',
+          type: 'SocketConnectionBox'
+        }
+      ],
+      javaCode: `
+      if ( box != null ) {
+        getBoxes().remove(box.getKey());
+
+        Socket socket = (Socket) box.getSocket();
+        if ( socket != null &&
+             socket.isConnected() ) {
+          try {
+            socket.getOutputStream().flush();
+          } catch (IOException e) {
+            // nop
+          }
+          try {
+            getLogger().debug("socket,close",socket.getRemoteSocketAddress());
+            socket.close();
+          } catch (IOException e) {
+            // nop
+          }
+        }
+      }
+      `
+    },
+    {
+      name: 'makeKey',
+      type: 'String',
+      args: [
+        {
+          name: 'host',
+          type: 'String'
+        },
+        {
+          name: 'port',
+          type: 'int'
+        }
+      ],
+      javaCode: `
+        return host + ":" + port;
+      `
+    },
+    {
+      name: 'get',
+      synchronized: true,
+      type: 'foam.box.Box',
+      args: [
+        {
+          name: 'x',
+          type: 'X'
+        },
+        {
+          name: 'host',
+          type: 'String'
+        },
+        {
+          name: 'port',
+          type: 'int'
+        }
+      ],
+      javaCode: `
+        String key = makeKey(host, port);
+        SocketConnectionBox box = (SocketConnectionBox) getBoxes().get(key);
+        if ( box != null ) {
+          return box;
+        }
+
+        try {
+          Socket socket = new Socket();
+          socket.setSoTimeout(getSoTimeout());
+          SocketAddress address = new InetSocketAddress(host, port);
+          socket.connect(address, getConnectTimeout());
+          box = new SocketConnectionBox(x, key, socket, host, port);
+          add(box);
```

Comment:
The 'boxes' collection is synchronized but there's no sync at this level to prevent two from being created. Is that going to be a problem?

Reply:
The model method is synchronized: 
`    {
      name: 'get',
      synchronized: true,
      type: 'foam.box.Box',
`